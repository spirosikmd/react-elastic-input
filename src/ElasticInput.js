import React from 'react';
import PropTypes from 'prop-types';

const TEXT_ATTRIBUTES = [
  'fontFamily',
  'fontSize',
  'fontWeight',
  'fontStyle',
  'letterSpacing',
  'textTransform',
  'wordSpacing',
];

const BORDER_BOX_ATTRIBUTES = [
  'paddingLeft',
  'paddingRight',
  'borderLeftStyle',
  'borderLeftWidth',
  'borderRightStyle',
  'borderRightWidth',
];

const PADDING_BOX_ATTRIBUTES = ['paddingLeft', 'paddingRight'];

class ElasticInput extends React.Component {
  constructor(props) {
    super(props);

    this.inputRef = React.createRef();
  }

  componentDidMount() {
    this.createWrapper();
    this.createMirror();

    this.updateInputWidth(
      this.inputRef.current,
      this.props.placeholder,
      this.mirror
    );
  }

  componentWillUnmount() {
    this.mirror.remove();
    this.wrapper.remove();
    cancelAnimationFrame(this.updateInputWidthFrame);
  }

  createWrapper() {
    this.wrapper = document.createElement('div');

    // Visually hide the wrapper element
    this.wrapper.style.position = 'fixed';
    this.wrapper.style.top = '-999px';
    this.wrapper.style.left = '0';

    document.body.appendChild(this.wrapper);
  }

  createMirror() {
    this.mirror = document.createElement('span');

    this.mirror.style.whiteSpace = 'pre';

    const inputComputedStyle = window.getComputedStyle(this.inputRef.current);
    this.setMirrorStyle(this.mirror, inputComputedStyle);

    this.wrapper.appendChild(this.mirror);
  }

  getWidthInPixels(value, mirror) {
    mirror.innerText = value;
    const delta = 1;
    return `${mirror.offsetWidth + delta}px`;
  }

  updateInputWidth(input, value, mirror) {
    this.updateInputWidthFrame = requestAnimationFrame(() => {
      input.style.width = this.getWidthInPixels(value, mirror);
    });
  }

  setMirrorStyle(mirror, inputComputedStyle) {
    TEXT_ATTRIBUTES.forEach((value) => {
      mirror.style[value] = inputComputedStyle[value];
    });

    mirror.style.paddingLeft = inputComputedStyle.textIndent;

    if (inputComputedStyle.boxSizing === 'border-box') {
      BORDER_BOX_ATTRIBUTES.forEach((value) => {
        mirror.style[value] = inputComputedStyle[value];
      });
    } else if (inputComputedStyle.boxSizing === 'padding-box') {
      PADDING_BOX_ATTRIBUTES.forEach((value) => {
        mirror.style[value] = inputComputedStyle[value];
      });
    }
  }

  handleInputChange = (event) => {
    const value = event.target.value || this.props.placeholder;
    this.updateInputWidth(this.inputRef.current, value, this.mirror);

    if (typeof this.props.onChange === 'function') {
      this.props.onChange(event);
    }
  };

  render() {
    /* eslint-disable no-unused-vars */
    const { type, onChange, ...props } = this.props;
    /* eslint-enable no-unused-vars */

    return (
      <input
        type={type}
        ref={this.inputRef}
        onChange={this.handleInputChange}
        {...props}
      />
    );
  }
}

ElasticInput.propTypes = {
  type: PropTypes.oneOf(['text', 'email', 'password', 'search', 'tel', 'url']),
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
};

ElasticInput.defaultProps = {
  type: 'text',
  placeholder: '',
  onChange: () => {},
};

export default ElasticInput;
