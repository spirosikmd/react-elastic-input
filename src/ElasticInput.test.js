import React from 'react';
import { mount } from 'enzyme';
import ElasticInput from './ElasticInput';

global.window.getComputedStyle = jest.fn().mockReturnValue({});

const div = global.document.createElement('div');
global.document.body.appendChild(div);

const renderComponent = (props) => {
  return mount(<ElasticInput {...props} />, { attachTo: div });
};

const testStyleProperties = (mirrorStyle, styleProperties) => {
  Object.keys(styleProperties).forEach((property) => {
    expect(mirrorStyle[property]).toBe(styleProperties[property]);
  });
};

describe('ElasticInput', () => {
  beforeEach(() => {
    jest
      .spyOn(global.window, 'requestAnimationFrame')
      .mockImplementation((cb) => cb());
  });

  afterEach(() => {
    global.window.requestAnimationFrame.mockRestore();
  });

  it('renders', () => {
    const component = renderComponent({ placeholder: 'placeholder' });
    expect(component).toMatchSnapshot();

    component.detach();
  });

  it('creates a wrapper element and attaches it to body', () => {
    const component = renderComponent();
    const instance = component.instance();

    expect(instance.wrapper).toBeDefined();
    expect(instance.wrapper.style.position).toBe('fixed');
    expect(instance.wrapper.style.top).toBe('-999px');
    expect(instance.wrapper.style.left).toBe('0px');

    // First child is the div with the input and second is the wrapper
    expect(document.body.childNodes.length).toBe(2);

    component.detach();
  });

  it('creates a mirror element and attaches it to wrapper', () => {
    global.window.getComputedStyle = jest.fn().mockReturnValue({
      textIndent: '50px',
    });
    const component = renderComponent();
    const instance = component.instance();

    expect(instance.mirror).toBeDefined();
    expect(instance.mirror.style.whiteSpace).toBe('pre');

    expect(instance.wrapper.childNodes.length).toBe(1);

    component.detach();
  });

  it('creates a mirror element and copies text styles from input', () => {
    const styleProperties = {
      fontFamily: 'Arial, Helvetica, sans-serif',
      fontSize: '15px',
      fontWeight: 'bold',
      fontStyle: 'italic',
      letterSpacing: '3px',
      textTransform: 'uppercase',
      wordSpacing: '30px',
    };

    global.window.getComputedStyle = jest.fn().mockReturnValue(styleProperties);
    const component = renderComponent();
    const instance = component.instance();

    testStyleProperties(instance.mirror.style, styleProperties);

    component.detach();
  });

  it('creates a mirror element and copies styles from input with border-box', () => {
    const styleProperties = {
      paddingLeft: '10px',
      paddingRight: '10px',
      borderLeftStyle: 'dotted',
      borderLeftWidth: 'thin',
      borderRightStyle: 'dotted',
      borderRightWidth: 'thin',
    };

    global.window.getComputedStyle = jest.fn().mockReturnValue({
      textIndent: '50px',
      boxSizing: 'border-box',
      ...styleProperties,
    });
    const component = renderComponent();
    const instance = component.instance();

    testStyleProperties(instance.mirror.style, styleProperties);

    component.detach();
  });

  it('creates a mirror element and copies styles from input with padding-box', () => {
    const styleProperties = {
      paddingLeft: '10px',
      paddingRight: '10px',
    };

    global.window.getComputedStyle = jest.fn().mockReturnValue({
      textIndent: '50px',
      boxSizing: 'padding-box',
      ...styleProperties,
    });
    const component = renderComponent();
    const instance = component.instance();

    testStyleProperties(instance.mirror.style, styleProperties);

    component.detach();
  });

  it('removes wrapper and mirror when it unmounts', () => {
    const component = renderComponent();
    const instance = component.instance();

    instance.componentWillUnmount();

    expect(document.body.childNodes.length).toBe(1);

    component.detach();
  });

  it('updates input width on change', () => {
    const onChangeSpy = jest.fn();
    const component = renderComponent({
      onChange: onChangeSpy,
    });

    const mockEvent = {
      target: {
        value: 'value',
      },
    };

    const input = component.find('input');
    input.simulate('change', mockEvent);
    expect(onChangeSpy).toHaveBeenCalled();

    expect(input.instance().style.width).toBe('1px');

    component.detach();
  });
});
