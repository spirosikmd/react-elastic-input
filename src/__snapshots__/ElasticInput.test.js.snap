// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticInput renders 1`] = `
<ElasticInput
  onChange={[Function]}
  placeholder="placeholder"
  type="text"
>
  <input
    onChange={[Function]}
    placeholder="placeholder"
    type="text"
  />
</ElasticInput>
`;
